### ISSUE TITLE
Link outdated in Plugin Manager, How to write a plugin

### ISSUE THREAD

**Issue by embix** [NuGetPackageExplorer/NuGetPackageExplorer#193]
Hi! Thanks for reporting this feature/bug/question! 

Please keep / fill in the relevant info from this template so that we can help you as best as possible.

**Type** (choose one): Bug
**NPE version**: 3.21.0.0

**Installed from**: ClickOnce

**current result** Opens NPE's codeplex page "How to: Write a plugin for NuGet Package Explorer" in browser. Which is for NPE 1 and 2. I am also aware that "CodePlex is shutting down" and "This project has moved".

**expected result** Opens browser with link to a github wiki page about "How to: Write a plugin for NuGet Package Explorer" for NPE 3.

**Known workarrounds:** None - I haven't found such a wiki page on NPE's github page so far
** Is there a version in which it did worked?** None that I know of.
**Is there a package on Nuget.org which shows the bug?** not applicable

**Comment by 304NotModified** [comment]
yes we need to port the page http://npe.codeplex.com/wikipage?title=How%20to%20write%20a%20plugin%20for%20NuGet%20Package%20Explorer indeed.

I think it still works, could you test that?

**Comment by embix** [comment]
It doesn't really work out:

I tried using the latest `NuGetPackageExplorer.Types` which is 4.0 on NuGet, 3.7.0.0 according to AssemblyVersion and AssemblyFileVersion. That version targets .net 4.5 so Step 1 of the How-To has to be updated.

Once that is done, one can walk the How-To up to step 5. Only, NPE doesn't require a bare assembly anymore, but a nuget package. Thankfully NPE let's you create one. NPE 3.22.0.0 (installed via Windows Store) tells you it expects the assembly in `lib\net40` although my package has to target 4.5 in order to reference the type assembly. If you then try to load your extension as nupkg via "Add Local Plugin", you'll get an error: _One or more errors occurred while loading the selected plugin: Could not load file or assembly 'NuGetPackageExplorer.Types, Version=3.7.0.0, Culture=neutral, PublicKeyToken=b33e3c5...'_

The 3.0 version of the type assembly has assembly version 3.0.0.0 and is targetting .net 4.0. Thus you get up to step 5 where you still have to provide a nuget package. Loading also fails, now with _One or more errors occurred while loading the selected plugin: Could not load file or assembly 'NuGetPackageExplorer.Types, Version=3.0.0.0, Culture=neutral, PublicKeyToken=b33e3c5...'_
_

HTH

BTW: Why aren't nuget version of the type package in sync with its assembly version?

**Comment by 304NotModified** [comment]
thx for the feedback. Currently NuGetPackageExplorer.Types isn't maintained (jumped here in without the info). Have to look at that.

**Comment by 304NotModified** [comment]
this seems the same prob. 
https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/issues/160

**Comment by 304NotModified** [comment]
[NuGetPackageExplorer.Types.5.0-alpha1.nupkg.zip](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/files/974774/NuGetPackageExplorer.Types.5.0-alpha1.nupkg.zip)

could you please try with the attached one?

**Comment by embix** [comment]
Earliest this evening but at the weekend is more likely.

**Comment by 304NotModified** [comment]
Sure no problem!

**Comment by embix** [comment]
[Unfortunately VS2017.1 doesn't let me add a local package source anymore][1], so I extracted the dll directly for my net4.5 project and used  IPackageContentViewer code usually generated on install and added the `System.ComponentModel.Composition` reference by hand. Build, package, load:

    One or more errors occurred while loading the selected plugin:
    
    Could not load file or assembly 'NuGetPackageExplorer.Types',
    Version=3.21.0.0, Culture=neutral, PublicKey=b33ec...
    or one of its dependencies. The system cannot find the file specified.

I am running NPE 3.22.0.0 here on my home machine, but reported the bug back at work. Sorry. Maybe a binding redirect can help?

 [1]: https://developercommunity.visualstudio.com/content/problem/52532/vs-just-crashed-while-trying-to-set-up-local-nuget.html

**Comment by 304NotModified** [comment]
> Sorry. Maybe a binding redirect can help?

Yes, please try. 

I will check the assembly version numbers. 

> Unfortunately VS2017.1 doesn't let me add a local package source anymore,

That really annoying

**Comment by 304NotModified** [comment]
FYI this is the project graph

![image](https://cloud.githubusercontent.com/assets/5808377/25762058/86f55f2c-31dd-11e7-906c-bc198888c5e3.png)

the attached nupkg was built with: (assembly version, not file version)

types: 3.21 (strong named)
core: 3.22 (not strong named)
packageexplorer: 3.22 (not strong named)
PackageViewModel: 3.22  (not strong named)
code executor: 1.0.0.0  (not strong named)

**Comment by 304NotModified** [comment]
this could be the issue: (in packageexplorer) (app.config)

```
    <dependentAssembly>
                <assemblyIdentity name="NuGetPackageExplorer.Types" publicKeyToken="b33e3c54f0b885e9" culture="neutral"/>
                <bindingRedirect oldVersion="1.0.0.0-3.10.0.0" newVersion="3.10.0.0"/>
            </dependentAssembly>
```

**Comment by embix** [comment]
Ok, at home I tried the Windows Store App which is 3.22 - I just deinstalled it and now VS accept's new nuget repos again. Must be an issue with the Windows Store App or Windows Store Policy. Anyway, I used your Types 5 branch and it finally worked out with just a binding redirect:

            <dependentAssembly>
                <assemblyIdentity name="NuGetPackageExplorer.Types" publicKeyToken="b33e3c54f0b885e9" culture="neutral"/>
                <bindingRedirect oldVersion="1.0.0.0-3.21.0.0" newVersion="3.21.0.0"/>
            </dependentAssembly>

I can provide a pull request if you want. One could argue for upwards compatibility and redirect to 3.9999 or so if desired. I'm just glad it works now.

**Comment by 304NotModified** [comment]
great news!

PR is great, but I'm not sure what is the best fix:

1. remove strong name of types (probably bad idea)
2. make the assembly of types fixed (e.g. 4.0.0.0)
3. document it really good when to change (and that it should be changed in app.config)

What do you think?

change it every version won't be an option, as your plugin won't work in newer version of NPE then.

**Comment by embix** [comment]
Interesting question. Not sure what our best option is.

 1. should not necessary in this case ([I just read this blog post][1]), we can use binding redirects
 2. if the API of the type assembly does not change, there is no need to change it's version. If we just add a type/method (non breaking) increasing the minor seems reasonable - along with that the binding redirect for the type library in the application has to be updated (max of old, fixed new version). If there are breaking changes, no binding redirect will help.
 3. At least comment in NuGetPackageExplorer/Types/Properties/AssemblyInfo.cs right above the AssemblyVersion attribute to update the PackageExplorer/app.config if you increment the version.

To form a better rooted opinion I'll have to fiddle around a bit

 [ ] I will branch v3.21 and test if an upward-compatible binding redirect works out. 
 [x] I just tested an old plugin (PackageDependencyVisualizer) from the feed and it didn't work out because types are missing: thus a breaking change, but not indicated in version as types 3.7.0.0 seems to be referenced (according to JustDecompile)
 [ ] I should try and create plugin based on an earlier version and use it current NPE.

 [1]: https://www.pedrolamas.com/2016/03/01/still-strong-naming-your-assemblies-you-do-know-its-2016-right/

**Comment by embix** [comment]
I just used my 3.21.0.0 targeting plugin in a "newer" NPE version using a 5.0.0.0 type library for testing. Plugin still worked. So I say we go the easiest way and just keep binding redirect and type library version in sync.

If we want to do right, we should also keep NuGet package version reflecting the dll version. To not break anything (4.0 is already public, 5.0-alpha in codebase) we could go with 5.0 for the next release.

**Comment by 304NotModified** [comment]
> If we want to do right, we should also keep NuGet package version reflecting the dll version. To not break anything (4.0 is already public, 5.0-alpha in codebase) we could go with 5.0 for the next release.

agree, if have created https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/pull/200. Should solve the issues.

**Comment by 304NotModified** [comment]
Note, this should be fixed. The docs aren't updated yet (those are here: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/wiki/How-to-Write-a-plugin-for-NuGet-Package-Explorer)

related: #741

**Comment by 304NotModified** [comment]
I was following https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/wiki/How-to-Write-a-plugin-for-NuGet-Package-Explorer to test it.

It works if we update the .Types package to 5.x and target netcoreapp3.0 

@onovotny Is there a reason .Types is not updated automatically? (and that 5.0.0 is unlisted?)

**Comment by onovotny** [comment]
There are a few issues around plugins so far:

1. We reference a custom build of NuGet. The version numbers may not match what's in public nuget.org for those same libraries. That may not matter though since the surface area is the same and the implementation we carry will "win."
2. Assembly loading is a bit different in Core. We'll probably need to use an AssemblyLoadContext, especially if a plugin has dependencies. Those plugins would need to be packages as the result of a publish (like a .net core global tool), so that they carry their dependencies locally. We'll need to come up with the right instructions for how to do that easily.
3. We need to think about where plugins should go. Do we load them from an arbitrary location and then copy them into the app area somewhere? The user won't know the app install location.

Overall, I hadn't given a lot of thought to plugins overall as demand for it wasn't really clear.

**Comment by 304NotModified** [comment]
> 2\. Assembly loading is a bit different in Core.

well it works, so not sure if we should change something about it?

**Comment by 304NotModified** [comment]
Update: the simple example works, trying now ILSpy which is a bit more complicated

**Comment by onovotny** [comment]
It works if there's a single file and there are no other dependencies whose version might conflict with a built-in one.

**Comment by 304NotModified** [comment]
got ILSply also working: https://github.com/NuGetPackageExplorer/Extensions.IlSpy/pull/2



> It works if there's a single file and there are no other dependencies whose version might conflict with a built-in one.

That was also before, isn't? Or was there an appdomain for it?
